### ISSUE TITLE
feat(data): synthetic data generation (EvolKit parity) and PII filtering

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#453]
## Summary

Add two data pipeline features to support the fine-tuning workflow:

## 1. Synthetic Data Generation (EvolKit Parity)

Automatic instruction complexity enhancement using self-play or an auxiliary
model. Generates harder training examples from seed instructions.

```bash
apr data evolve \
  --seed-data initial-instructions.jsonl \
  --model teacher-model.apr \
  --evolution-rounds 3 \
  --output evolved-data.jsonl
```

**Evolution strategies**:
- Complexity enhancement (add constraints, multi-step reasoning)
- Breadth enhancement (vary domains, contexts)
- Concretization (make abstract instructions specific)

**Reference**: Based on Microsoft's "Automatic Instruction Evolving" paper,
implemented by Arcee's EvolKit (open source).

## 2. PII Filtering

Automatic detection and removal of PII from training data before fine-tuning.

```bash
apr data filter-pii --input raw-data.jsonl --output clean-data.jsonl
```

**Detected PII types**: email, phone, SSN, credit card, IP address, names
(regex + NER-based patterns).

## 3. Domain-Specific Data Filtering

Quality scoring and filtering of training data based on domain relevance.

```bash
apr data filter --input raw.jsonl --domain legal --min-quality 0.7 -o filtered.jsonl
```

## Specification

See `docs/specifications/components/finetune.md` §6.1-6.3.

### Contract: `data-pipeline-v1.yaml`

```yaml
metadata:
  description: "Data quality pipeline — synthetic data generation and PII filtering"
  references:
    - "Xu et al. (2023) WizardLM: Automatic Instruction Evolving"

equations:
  pii_filter_completeness:
    formula: "∀ record r: pii_scan(r) → {type, span, confidence}"
    invariants:
      - "Known PII patterns never pass through filter"
      - "Non-PII text is preserved unchanged"
      - "Filter is idempotent: filter(filter(x)) == filter(x)"

  evolution_diversity:
    formula: "diversity(evolved) >= diversity(seed)"
    invariants:
      - "Evolved instructions are semantically distinct from seeds"
      - "Output count >= input count (no data loss)"
      - "JSON format preserved through evolution"

proof_obligations:
  - type: invariant
    property: "PII filter idempotency"
    formal: "filter(filter(data)) == filter(data)"
  - type: invariant
    property: "Non-PII preservation"
    formal: "text without PII patterns passes through unchanged"
  - type: completeness
    property: "Known PII detection"
    formal: "email, phone, SSN regex patterns always detected"
  - type: invariant
    property: "JSONL format preservation"
    formal: "output is valid JSONL if input is valid JSONL"
  - type: bound
    property: "No data loss in evolution"
    formal: "len(evolved) >= len(seed)"

falsification_tests:
  - id: FALSIFY-PII-001
    rule: "Email detection"
    prediction: "user@example.com always detected and redacted"
    if_fails: "Email regex pattern missing or incorrect"
  - id: FALSIFY-PII-002
    rule: "Idempotency"
    prediction: "double-filtering produces same output as single filter"
    if_fails: "Filter creates new PII-like patterns"
  - id: FALSIFY-PII-003
    rule: "Non-PII preservation"
    prediction: "text without PII unchanged after filtering"
    if_fails: "Overly aggressive PII regex"
  - id: FALSIFY-EVOL-001
    rule: "Format preservation"
    prediction: "evolved output is valid JSONL"
    if_fails: "Evolution corrupts JSON structure"
  - id: FALSIFY-EVOL-002
    rule: "No data loss"
    prediction: "output record count >= input record count"
    if_fails: "Evolution silently drops records"
```

### Binding Requirements

Add to `contracts/aprender/binding.yaml`:
```yaml
  - contract: data-pipeline-v1.yaml
    equation: pii_filter_completeness
    module_path: "aprender::data::pii"
    function: filter_pii
    status: not_implemented

  - contract: data-pipeline-v1.yaml
    equation: evolution_diversity
    module_path: "aprender::data::evolve"
    function: evolve_instructions
    status: not_implemented
```

### Verification Targets
- L2: 5 falsification tests
- L3: Kani harness for PII regex completeness (bounded input ≤ 64 chars)
